tests/func/test_scm.py

```diff
@@ -99,6 +99,21 @@ def test_ignored(tmp_dir, scm):
     assert not scm.is_ignored(tmp_dir / "dir1" / "file2.txt")
 
 
+def test_ignored_dir_unignored_subdirs(tmp_dir, scm):
+    tmp_dir.gen({".gitignore": "data/**\n!data/**/\n!data/**/*.csv"})
+    scm.add([".gitignore"])
+    tmp_dir.gen(
+        {"data/raw/tracked.csv": "cont", "data/raw/not_tracked.json": "cont"}
+    )
+
+    assert not scm.is_ignored("data/raw/tracked.csv")
+    assert scm.is_ignored("data/raw/not_tracked.json")
+
+    assert not scm.is_ignored("data/")
```

Comment:
I was considering adding also `assert scm.is_ignored("data")`.
It passes, but it's kind of unclear to me what the outcome should be there - is it an universally correct behavior or just  a `dulwich` interpretation. 
I'd prefer not to add it after all

Reply:
Testing it with CLI `git check-ignore` looks like it is supposed to pass, so it looks like dulwich's behavior is correct

As far as I can tell CLI git only ignores it if you explicitly include the trailing `/`

Reply:
So in this case, CLI check-ignore is doing the `isdir` check on the supplied local filesystem path, and then compares it to the pattern with trailing slash if it is a directory.

But in our case, in `scm.is_ignored()` we are already assuming that the supplied path has had the trailing slash explicitly added if `path` is supposed to be a directory.

So in this check:
```
assert scm.is_ignored("data/raw")
```
we are explicitly checking the file path `data/raw` and not the directory path `data/raw/`. Meaning the assertion should pass.

It would be good to add a comment about this behavior in the test regarding why it will differ from CLI `git check-ignore` though.

---

And actually, for this test you shouldn't need to generate the file content at all (lines 105-107), as `scm.is_ignored` only deals with checking raw paths against the ignore patterns, and does not care whether or not the paths exist (or are files/dirs).

Reply:
also, note that you can get the same behavior with CLI `git check-ignore` if you are comparing raw paths that don't exist in your local working directory

```
ignore-test git:master ✩   ❯ cat .gitignore
dir/**
!dir/**/
!dir/**/*.dvc

ignore-test git:master ✩   ❯ ls

ignore-test git:master ✩   ❯ git check-ignore dir

ignore-test git:master ✩   ❯ git check-ignore dir/                                                                                    ⏎
dir/

ignore-test git:master ✩   ❯ git check-ignore dir/subdir
dir/subdir

ignore-test git:master ✩   ❯ git check-ignore dir/subdir/
dir/subdir/

ignore-test git:master ✩   ❯ git check-ignore dir/subdir/foo.dvc

ignore-test git:master ✩   ❯ git check-ignore dir/subdir/foo                                                                          ⏎
dir/subdir/foo
```

although I'm not entirely sure why `dir/subdir/` doesn't pass here, as my understanding was that it should be excluded by `!dir/**/`

Reply:
Perhaps this is because of the `is_dir` check?  (since `dir/subdir/` is not an existing directory, perhaps the trailing slash gets stripped internally?)

In any case, do you think it would be okay to merge this as is? My reasoning is: let's test the behavior we actually need (focusing on existing files and dirs) and not all of the `dulwich`/`git check-ignore` contract, as that might be a huge task

Reply:
The issue is that checking ignores against raw paths which don't exist locally is a needed use-case in DVC, as when we do operations with a specific git-revision (like `dvc push --rev=...`), we will be checking raw paths (that might not exist at all locally) directly against git trees.

So handling git-ignored/excluded `.dvc` files consistently is a requirement, whether we are pushing what's in the user's local workspace, or what is in some other git revision/branch (that is not checked-out in the git working tree)

Reply:
Looking at the CGit source, they do only treat paths passed into `check-ignore` as directories if the path actually exists, and is a directory, and will end up removing the trailing slash from the path if the is_directory check fails before checking against the gitignore patterns

the paths are checked using this `is_directory()` function:
https://github.com/git/git/blob/142430338477d9d1bb25be66267225fb58498d92/abspath.c#L8

which confirms what you were thinking about the `git check-ignore` behavior

---

So at a minimum, given the gitignore patterns

```
dir/**
!dir/**/
!dir/**/*.dvc
```

I think we should be testing with the following raw paths:

```
assert not scm.is_ignored("dir")
assert scm.is_ignored("dir/file")
assert not scm.is_ignored("dir/subdir/")
assert scm.is_ignored("dir/subdir/file")
assert not scm.is_ignored("dir/subdir/file.dvc")
---

Reply:
I see. Didn't think about that. Added a comment as I understand it - and assertions for both existing and made-up paths

Reply:
Only saw your second comment after submitting.  Sorry about the commit spam :wink: 

Adding a few more assertions

Reply:
Thanks! It looks like this PR should be good now as long as CI passes